signac/__main__.py

```diff
@@ -1045,10 +1045,13 @@ def jobs():
                     try:
                         readline.read_history_file(fn_hist)
                         readline.set_history_length(1000)
-                    except (IOError, OSError) as error:
-                        if error.errno != errno.ENOENT:
-                            raise
-                    atexit.register(readline.write_history_file, fn_hist)
+                        with open(fn_hist, 'w') as f: pass
+                        atexit.register(readline.write_history_file, fn_hist)
```

Comment:
Should we maybe just open it in append (`a`) mode so that in case that we fail to write the history from memory at program exit it is at least not lost?

Reply:
@csadorf yes seems like a good idea. done that!

Reply:
@csadorf I'm not sure if I understand. How does using `a` fix the problem if we don't have write access? I thought the primary difference between `w` and `a` was whether it overwrote the existing file or appended to the end.

Reply:
@bdice actually the main reason for line-1048 is to check weather the file has a write permission. If it doesn't then we get an exception that is used to throw a warning.

Reply:
@ac-optimus @csadorf I would prefer to use the approach of permissions-checking for this specific instance, instead of attempting to write to the file and catching an exception. I believe that opening the file (using `with open(...): pass`) requires more system calls and may have unintended side-effects such as memory allocation (for the file buffer) and updating the "modified" timestamp of the file with no actual modifications. My preferred solution would be to use `os.access` as before.

Reply:
The [documentation for `os.access`](https://docs.python.org/3/library/os.html#os.access) explicitly recommends to use the [EAFP](https://docs.python.org/3/glossary.html#term-eafp) approach instead. The exact motivation for this (security issue) might not apply here, but your reasoning does not make any sense to me either, since this particular code segment is only executed once per opening of signac shell. The number of system calls should be completely irrelevant. I also don't see any issue with updating the modification time stamp.

IMO, the approach that we should take here, should be the one that is most robust across platforms. With that in mind, attempting to open the file for writing is the most surefire way to figure out whether we have the rights to do that.

The only reason I recommended 'a' over 'w' is that with the former we don't overwrite the file.

Reply:
@bdice @csadorf what to do now?

Reply:
@ac-optimus A few notes:

- Using mode `'a'` would definitely be correct, or else the file will be overwritten with empty contents when we check if it can be opened. I apologize, I should have understood that earlier.
- I agree with @csadorf in general, that EAFP is preferred. However, I think we're going about this in the wrong way. The file is written _at exit time_, and we're checking whether it's writable _during shell launch_. No matter whether you use `os.access` or attempt to write using `with open(...)`, the file permissions could be different at exit time. I believe the best approach would be to create a function `write_history_file` that attempts to save the file and prints a warning if the history could not be saved, and register `write_history_file` to be run at exit time.

My suggestion for this block would be the following:
```python
                    try:
                        readline.read_history_file(fn_hist)
                        readline.set_history_length(1000)
                    except FileNotFoundError:
                        pass
                    except PermissionError:
                        print("Warning: Shell history could not be read from "
                              "{}.".format(os.path.relpath(fn_hist)))
                    def write_history_file():
                        try:
                            readline.write_history_file(fn_hist)
                        except PermissionError:
                            print("Warning: Shell history could not be written to "
                                  "{}.".format(os.path.relpath(fn_hist)))
                    atexit.register(write_history_file)
```

Reply:
@bdice  seems a good fix but I want to point to something.  `readline.write_history_file(fn_hist)` does not throw `PermissionError` when we don't have write access but rather changes the permission of the file. And hence whatever be the        case( read/write, read-only, write-only, no read and write) the premission changes once `readline.write_history_file(fn_hist)`  executes. Because of this I added `atexit.register(readline.write_history_file, fn_hist)` when we have write permission and checked the write access using ```with``` statement.
```python
try:
    readline.read_history_file(fn_hist)
    readline.set_history_length(1000)
    with open(fn_hist, 'w'):
        pass
    atexit.register(readline.write_history_file, fn_hist)
except FileNotFoundError:
    atexit.register(readline.write_history_file, fn_hist)
except PermissionError:
    print(("Warning: {} does not have read/write permission. The history "
            "of this shell will not be saved.").format(os.path.relpath(fn_hist)))

```

Reply:
@ac-optimus The original issue #262 was about reading the file. If writing the file succeeds and changes the permission of the file, that is fine. I believe that `readline.write_history_file` may be deleting the current file and re-creating it. In my understanding, deleting a file is allowed if the user has write permissions on its containing directory. You may need to see if a `PermissionError` is raise if the user does not have write permission on the containing directory.

Reply:
I thought we don't have to disturb the permission of the file. So in that case if we have read and write permission then we append to .signac_shell_history else we delete the current file and create new one to save the current shell history. Because``atexit.register(readline.write_history_file, fn_hist)`` will anyways create a new file if either read or write permission is not there.
And ``PermissionError`` will not be raised  as ``readline.write_history_file`` re-creates the file, when no write or read permission is there. So we might have to add ``with open(...)`` to check the file permission. And even warning might change. Please correct me if I am wrong.
The modified code, but here the write warning will appear after the shell exits --
```python
                     try:
                        readline.read_history_file(fn_hist)
                        readline.set_history_length(1000)
                    except FileNotFoundError:
                        pass
                    except PermissionError:
                        print("Warning: Shell history from "
                              "{}.".format(os.path.relpath(fn_hist)),
                              " would be lost. No read permission.")
                    def write_history_file():
                        try:
                            with open(fn_hist, 'a'):
                                pass
                        except PermissionError:
                            print("Warning: Shell history from "
                                  "{}.".format(os.path.relpath(fn_hist)),
                                  " would be lost. No write permission.")
                        readline.write_history_file(fn_hist)
                    atexit.register(write_history_file)
```
If we want the write warning to show up the moment the shell starts like the read warning then the code would be -
```python

                    try:
                        readline.read_history_file(fn_hist)
                        readline.set_history_length(1000)
                    except FileNotFoundError:
                        pass
                    except PermissionError:
                        print("Warning: Shell history from "
                              "{}.".format(os.path.relpath(fn_hist)),
                              " would be lost. No read premission.")
                    try:
                        with open(fn_hist, 'a'):
                             pass
                    except PermissionError:
                        print("Warning: Shell history from "
                              "{}.".format(os.path.relpath(fn_hist)),
                              " would be lost. No write permission.")
                    atexit.register(readline.write_history_file, fn_hist)

```

Reply:
or we can throw common error for both read and write permission like this-
```python
                    try:
                        readline.read_history_file(fn_hist)
                        readline.set_history_length(1000)
                        with open(fn_hist, "a"):
                            pass
                    except FileNotFoundError:
                        pass
                    except PermissionError:
                        print("Warning: Shell history from "
                              "{}".format(os.path.relpath(fn_hist)),
                              "would be lost. Permission error.")
                    atexit.register(readline.write_history_file, fn_hist)


```

Reply:
I'd just go with the last alternative, since I believe that it solves the original issue. @bdice Do you agree?

Reply:
I recommend that we adopt the snippet I proposed above: https://github.com/glotzerlab/signac/pull/279#discussion_r373607211
@ac-optimus, can you apply that snippet and test it out manually? (no need to write a unit test)

Reply:
@bdice Sounds good to me!

Reply:
@bdice I already tried it manually and it was the reason why I proposed slight modification (1st code spinet above). ```readline.write_history_file(fn_hist)``` does not throw error when we do not have write permission and hence the exception is never called. And the 3rd one is more general. Please have a look and correct me if I am wrong.